File: Object-C/BlackFireIM/Example/Home/Controller/BFHomeController.m
//
//  BFHomeController.m
//  BlackFireIM
//
//  Created by benny wang on 2021/4/8.
//

#import "BFHomeController.h"
#import "MSIMSDK-UIKit.h"
#import <MSIMSDK/MSIMSDK.h>
#import "BFSparkCardView.h"
#import "BFSparkCardCell.h"
#import "BFChatViewController.h"
#import "BFSparkLoadingView.h"
#import "BFSparkEmptyView.h"
#import "BFAnimationView.h"


@interface BFHomeController()<BFSparkCardViewDelegate,BFSparkCardViewDataSource,BFSparkCardCellDelegate>

@property(nonatomic,strong) BFSparkCardView *containter;

@property(nonatomic,strong) BFSparkLoadingView *loadingView;

@property(nonatomic,strong) BFSparkEmptyView *emptyView;

@property(nonatomic,strong) UIButton *likeBtn;

@property(nonatomic,strong) UIButton *dislikeBtn;

@property(nonatomic,strong) NSMutableArray *dataList;

@end
@implementation BFHomeController

- (void)viewDidLoad
{
    [super viewDidLoad];
    [self setupUI];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self loadParksData];
    });
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self.navigationController setNavigationBarHidden:YES animated:animated];
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    [self.navigationController setNavigationBarHidden:NO animated:animated];
}

- (BFSparkLoadingView *)loadingView
{
    if (!_loadingView) {
        _loadingView = [[BFSparkLoadingView alloc]init];
    }
    return _loadingView;
}

- (void)setupUI
{
    self.navView.hidden = YES;
//    self.likeBtn = [UIButton buttonWithType:UIButtonTypeCustom];
//    [self.likeBtn setImage:[UIImage imageNamed:@"card_like"] forState:UIControlStateNormal];
//    [self.likeBtn addTarget:self action:@selector(likeAction:) forControlEvents:UIControlEventTouchUpInside];
//    self.likeBtn.frame = CGRectMake(Screen_Width*0.5+30, Screen_Height-TabBar_Height-15-60, 60, 60);
//    self.likeBtn.hidden = YES;
//    [self.view addSubview:self.likeBtn];
//
//    self.dislikeBtn = [UIButton buttonWithType:UIButtonTypeCustom];
//    [self.dislikeBtn setImage:[UIImage imageNamed:@"card_dislike"] forState:UIControlStateNormal];
//    [self.dislikeBtn addTarget:self action:@selector(dislikeAction:) forControlEvents:UIControlEventTouchUpInside];
//    self.dislikeBtn.frame = CGRectMake(Screen_Width*0.5-30-60, Screen_Height-TabBar_Height-15-60, 60, 60);
//    self.dislikeBtn.hidden = YES;
//    [self.view addSubview:self.dislikeBtn];
    
    self.loadingView.frame = CGRectMake(0, 0, Screen_Width, Screen_Height);
    [self.view addSubview:self.loadingView];
    [self.loadingView beginAnimating];
    
    CGFloat maxH = Screen_Height-StatusBar_Height-10-TabBar_Height-10;
    CGFloat cardW = Screen_Width-30;
    CGFloat cardH = MIN(cardW/0.6, maxH);
    self.containter = [[BFSparkCardView alloc]initWithFrame:CGRectMake(15, StatusBar_Height + 10 + (maxH-cardH)*0.5, cardW, cardH)];
    self.containter.delegate = self;
    self.containter.dataSource = self;
    self.containter.visibleCount = 3;
    self.containter.lineSpacing = 10;
    self.containter.interitemSpacing = 10;
    self.containter.maxAngle = 15;
    self.containter.maxRemoveDistance = 100;
    [self.containter registerClass:[BFSparkCardCell class] forCellReuseIdentifier:@"cardCell"];
    [self.view addSubview:self.containter];
    
    self.emptyView = [[BFSparkEmptyView alloc]initWithFrame:CGRectMake(0, Screen_Height*0.35, Screen_Width, 235)];
    self.emptyView.hidden = YES;
    [self.emptyView.retryBtn addTarget:self action:@selector(retryButtonClick) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:self.emptyView];
    
    self.dataList = [NSMutableArray array];
}

- (void)loadParksData
{
    [[MSIMManager sharedInstance] getSparks:^(NSArray<MSProfileInfo *> * sparks) {
        [self.dataList removeAllObjects];
        [self.dataList addObjectsFromArray:sparks];
        [self bf_reloadData];
        } fail:^(NSInteger code, NSString * _Nonnull desc) {
            [MSHelper showToastFail:desc];
            [self.loadingView stopAnimating];
            self.emptyView.hidden = NO;
    }];
}

- (void)bf_reloadData
{
    self.containter.alpha = 0;
    self.likeBtn.alpha = 0;
    self.dislikeBtn.alpha = 0;
    self.emptyView.hidden = YES;
    [UIView animateWithDuration:0.5 animations:^{
        self.containter.alpha = 1;
        self.likeBtn.alpha = 1;
        self.dislikeBtn.alpha = 1;
        
    } completion:^(BOOL finished) {
        [self.loadingView stopAnimating];
        self.emptyView.hidden = self.dataList.count != 0;
    }];
    [self.containter reloadData];
}

- (void)retryButtonClick
{
    [self loadParksData];
}

#pragma mark -- BFSparkCardViewDelegate,BFSparkCardViewDataSource
- (NSInteger)numberOfCountInCardView:(BFSparkCardView *)cardView
{
    return self.dataList.count;
}

- (BFCardViewCell *)cardView:(BFSparkCardView *)cardView cellForRowAtIndex:(NSInteger)index
{
    BFSparkCardCell *cell = [cardView dequeueReusableCellWithIdentifier:@"cardCell"];
    [cell configItem:self.dataList[index]];
    cell.delegate = self;
    return cell;
}

- (void)cardView:(BFSparkCardView *)cardView didRemoveCell:(BFSparkCardCell *)cell forRowAtIndex:(NSInteger)index direction:(CardCellSwipeDirection)direction
{
    cell.like.alpha = 0;
    cell.dislike.alpha = 0;
}

- (void)cardView:(BFSparkCardView *)cardView didRemoveLastCell:(BFSparkCardCell *)cell forRowAtIndex:(NSInteger)index
{
    self.emptyView.hidden = NO;
}

- (void)cardView:(BFSparkCardView *)cardView didDisplayCell:(BFSparkCardCell *)cell forRowAtIndex:(NSInteger)index
{
    
}

- (void)cardView:(BFSparkCardView *)cardView didMoveCell:(BFSparkCardCell *)cell forMovePoint:(CGPoint)point direction:(CardCellSwipeDirection)direction
{
    if (direction == CardCellSwipeDirectionLeft) {
        cell.like.alpha = 0.0f;
        cell.dislike.alpha = 1.0f;
        cell.dislike.transform = CGAffineTransformMakeRotation(45*M_PI / 180.0);
    }else if (direction == CardCellSwipeDirectionRight) {
        cell.dislike.alpha = 0.0f;
        cell.like.alpha = 1.0f;
        cell.like.transform = CGAffineTransformMakeRotation(-45*M_PI / 180.0);
    }else {
        cell.like.alpha = 0.0f;
        cell.dislike.alpha = 0.0f;
    }
}


#pragma mark - Action Methods
- (void)dislikeAction:(UIButton *)sender
{
    [self.containter removeTopCardViewFromSwipe:CardCellSwipeDirectionLeft];
}

- (void)likeAction:(UIButton *)sender
{
    [self.containter removeTopCardViewFromSwipe:CardCellSwipeDirectionRight];
}

- (void)winkBtnDidClick:(BFSparkCardCell *)cell
{
    if (cell.user.user_id && cell.winkBtn.isSelected == NO) {
        MSIMEmotionElem *elem = [[MSIMEmotionElem alloc]init];
        elem.emotionID = @"001";
        elem.emotionName = @"emotion_01";
        elem = [[MSIMManager sharedInstance]createEmotionMessage:elem];
        [[MSIMManager sharedInstance]sendC2CMessage:elem toReciever:cell.user.user_id successed:^(NSInteger msg_id) {
                    
             cell.winkBtn.selected = YES;
            [BFAnimationView showAnimation:@"spark_like" size:CGSizeMake(200, 200) isLoop:NO];
            
            } failed:^(NSInteger code, NSString * _Nonnull desc) {
                [MSHelper showToastFail:desc];
        }];
    }
}

- (void)chatBtnDidClick:(BFSparkCardCell *)cell
{
    if (cell.user.user_id) {
        BFChatViewController *vc = [[BFChatViewController alloc]init];
        vc.partner_id = cell.user.user_id;
        [self.navigationController pushViewController:vc animated:YES];
    }
}


@end

File: Swift/MSIM-Swift/Home/BFHomeController.swift
//
//  BFHomeController.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/8/2.
//

import UIKit
import MSIMSDK



class BFHomeController: BFBaseViewController, BFSparkCardCellDelegate {
    

    lazy var container: BFSparkCardView = {
        
        let maxH = UIScreen.height - UIScreen.statusbarHeight - 10 - UIScreen.tabBarHeight - 10;
        let cardW = UIScreen.width - 30;
        let cardH = min(cardW / 0.6, maxH)
        let container = BFSparkCardView(frame: CGRect(x: 15, y: UIScreen.statusbarHeight + 10 + (maxH - cardH) * 0.5, width: cardW, height: cardH))
        container.delegate = self
        container.dataSource = self
        container.visibleCount = 3
        container.lineSpacing = 10
        container.interitemSpacing = 10
        container.maxAngle = 15
        container.maxRemoveDistance = 100
        container.registerClass(cellClass: BFSparkCardCell.self, forCellReuseIdentifier: "cardCell")
        return container
    }()
    
    lazy var loadingView: BFSparkLoadingView = {
        let loadingView = BFSparkLoadingView(frame: CGRect(x: 0, y: 0, width: UIScreen.width, height: UIScreen.height))
        return loadingView
    }()
    
    lazy var emptyView: BFSparkEmptyView = {
        let emptyView = BFSparkEmptyView(frame: CGRect(x: 0, y: UIScreen.height * 0.35, width: UIScreen.width, height: 235))
        emptyView.isHidden = true
        emptyView.retryBtn.addTarget(self, action: #selector(retryButtonClick), for: .touchUpInside)
        return emptyView
    }()
    
    var dataList: [MSProfileInfo] = []
    
    override func viewDidLoad() {
        super.viewDidLoad()
        view.addSubview(loadingView)
        loadingView.beginAnimating()
        
        view.addSubview(container)
        view.addSubview(emptyView)
        DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
            self.loadParksData()
        }
    }

    func loadParksData() {
        MSIMManager.sharedInstance().getSparks { sparks in
            
            self.dataList.removeAll()
            self.dataList += sparks
            self.bf_reloadData()
            
        } fail: { code, desc in
            
            MSHelper.showToastFailWithText(text: desc ?? "")
            self.loadingView.stopAnimating()
            self.emptyView.isHidden = false
        }
    }
    
    func bf_reloadData() {
        
        container.alpha = 0
        emptyView.isHidden = true
        UIView.animate(withDuration: 0.5) {
            self.container.alpha = 1
    
        } completion: { _ in
            self.loadingView.stopAnimating()
            self.emptyView.isHidden = self.dataList.count != 0
        }
        container.reloadData()
    }
    
    @objc func retryButtonClick() {
        loadParksData()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        navigationController?.setNavigationBarHidden(true, animated: animated)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        navigationController?.setNavigationBarHidden(false, animated: animated)
    }

}

extension BFHomeController: BFSparkCardViewDelegate,BFSparkCardViewDataSource {
    func cardView(cardView: BFSparkCardView, didRemoveCell: BFCardViewCell, forRowAtIndex: Int, direction: CardCellSwipeDirection) {
        let cell = didRemoveCell as! BFSparkCardCell
        cell.like.alpha = 0
        cell.dislike.alpha = 0
    }
    
    func cardView(cardView: BFSparkCardView, didRemoveLastCell: BFCardViewCell, forRowAtIndex: Int) {
        emptyView.isHidden = false
    }
    
    func cardView(cardView: BFSparkCardView, didDisplayCell: BFCardViewCell, forRowAtIndex: Int) {
        
    }
    
    func cardView(cardView: BFSparkCardView, didMoveCell: BFCardViewCell, forMovePoint: CGPoint, direction: CardCellSwipeDirection) {
        let cell = didMoveCell as! BFSparkCardCell
        if direction == .left {
            cell.like.alpha = 0
            cell.dislike.alpha = 1
            cell.dislike.transform = CGAffineTransform(rotationAngle: CGFloat(45 * Double.pi) / 180)
        }else if direction == .right {
            cell.dislike.alpha = 0
            cell.like.alpha = 1
            cell.like.transform = CGAffineTransform(rotationAngle: CGFloat(-45 * Double.pi) / 180)
        }else {
            cell.like.alpha = 0
            cell.dislike.alpha = 0
        }
    }
    
    
    func numberOfCountInCardView(cardView: BFSparkCardView) -> Int {
        return dataList.count
    }
    
    func cellForRowAtIndex(cardView: BFSparkCardView, index: Int) -> BFCardViewCell {
        let cell = cardView.dequeueReusableCellWithIdentifier(identifier: "cardCell") as! BFSparkCardCell
        cell.configItem(item: dataList[index])
        cell.delegate = self
        return cell
    }
    
    func winkBtnDidClick(cell: BFSparkCardCell) {
        if cell.profile?.user_id != nil && cell.winkBtn.isSelected == false {
            let elem = MSIMEmotionElem()
            elem.emotionID = "008"
            elem.emotionName = "emotion_08"
            let message = MSIMManager.sharedInstance().createEmotionMessage(elem)
            MSIMManager.sharedInstance().sendC2CMessage(message, toReciever: cell.profile!.user_id) { msg_id in
                
                cell.winkBtn.isSelected = true
            } failed: { code, desc in
                MSHelper.showToastFailWithText(text: desc ?? "")
            }

        }
    }
    
    func chatBtnDidClick(cell: BFSparkCardCell) {
        if let uid = cell.profile?.user_id {
            let vc = BFChatViewController()
            vc.partner_id = uid
            navigationController?.pushViewController(vc, animated: true)
        }
    }
}
